### PR TITLE
refactor(core): start refactoring pbs code

### PR DESCRIPTION
**Pull request by sarah-ek**
closes: _please link all relevant issues_

### Check-list:

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: tfhe/src/core_crypto/algorithms/ggsw_conversion.rs [hunk 1]
```diff
@@ -3,14 +3,14 @@
 //! like the Fourier domain.
 
 use crate::core_crypto::commons::computation_buffers::ComputationBuffers;
+use crate::core_crypto::commons::math::fft::{Fft, FftView};
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/ggsw_conversion.rs:6]: should we maybe plan for the fft64/fft128 split as we had in the fft_impl module ? I believe we want to also move the f128 at some point, we can check with JB

File: tfhe/src/core_crypto/algorithms/lwe_bootstrap_key_conversion.rs [hunk 2]
```diff
@@ -2,17 +2,58 @@
...
 use crate::core_crypto::fft_impl::fft128::math::fft::Fft128;
-use crate::core_crypto::fft_impl::fft64::crypto::bootstrap::{
-    fill_with_forward_fourier_scratch, FourierLweBootstrapKey,
-};
-use crate::core_crypto::fft_impl::fft64::math::fft::{Fft, FftView};
 use concrete_fft::c64;
-use dyn_stack::{PodStack, SizeOverflow, StackReq};
+use dyn_stack::{PodStack, ReborrowMut, SizeOverflow, StackReq};
+
+impl<'a> FourierLweBootstrapKeyMutView<'a> {
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_bootstrap_key_conversion.rs:21]: should this be in the entities module for now ?

My reasoning : the algorithm for now is tied to the entity, what we would do is the same you did in the ggsw_conversion module i.e. rewrite the algorithm in this file as a function not in an impl block ?

File: tfhe/src/core_crypto/commons/math/fft/mod.rs [hunk 3]
```diff
@@ -738,24 +711,23 @@ pub fn par_convert_polynomials_list_to_fourier<Scalar: UnsignedTorus>(
...
             for (fourier_poly, standard_poly) in izip!(
                 fourier_poly_chunk.chunks_exact_mut(f_polynomial_size),
                 standard_poly_chunk.chunks_exact(polynomial_size.0)
             ) {
                 fft.forward_as_torus(
-                    FourierPolynomialMutView { data: fourier_poly },
+                    FourierPolynomialMutView::from_container(fourier_poly),
                     PolynomialView::from_container(standard_poly),
                     stack.rb_mut(),
                 );
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/commons/math/fft/mod.rs:733]: could we put the mod tests that is at the the end of the file towards the top ?

File: tfhe/src/core_crypto/fft_impl/fft64/math/mod.rs [hunk 4]
```diff
@@ -1,3 +0,0 @@
-pub mod decomposition;
-pub mod fft;
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/fft_impl/fft64/math/mod.rs:2]: looks like fft has not been deleted ?

File: tfhe/src/core_crypto/entities/fourier_ggsw_ciphertext.rs [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+    pub fn new(
+        data: C,
+        glwe_size: GlweSize,
+        polynomial_size: PolynomialSize,
+        row_count: usize,
+        decomposition_level: DecompositionLevel,
+    ) -> Self {
+        assert_eq!(
+            data.container_len(),
+            polynomial_size.to_fourier_polynomial_size().0 * glwe_size.0 * row_count
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/entities/fourier_ggsw_ciphertext.rs:137]: row_count should be equal to glwe_size you can check ggsw_ciphertext.rs I think there are some size functions :)

**Review by IceTDrinker — COMMENTED** [review]
First batch of comments, thanks for the work, I'll check the remainder after lunch :)

File: tfhe/src/core_crypto/entities/fourier_lwe_bootstrap_key.rs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    pub fn decomposition_base_log(&self) -> DecompositionBaseLog {
+        self.decomposition_base_log
+    }
+
+    pub fn decomposition_level_count(&self) -> DecompositionLevelCount {
+        self.decomposition_level_count
+    }
+
+    pub fn output_lwe_dimension(&self) -> LweDimension {
+        LweDimension((self.glwe_size.0 - 1) * self.polynomial_size().0)
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/entities/fourier_lwe_bootstrap_key.rs:94]: you can now use self.glwe_size.to_glwe_dimension().to_equivalent_lwe_dimension(self.polynomial_size())

File: tfhe/src/core_crypto/entities/fourier_ggsw_ciphertext.rs [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+    decomposition_level_count: DecompositionLevelCount,
+}
+
+/// A matrix containing a single level of gadget decomposition, in the Fourier domain.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct FourierGgswLevelMatrix<C: Container<Element = c64>> {
+    data: C,
+    glwe_size: GlweSize,
+    polynomial_size: PolynomialSize,
+    row_count: usize,
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/entities/fourier_ggsw_ciphertext.rs:27]: should not be needed row_count == glwe_size

File: tfhe/src/core_crypto/algorithms/lwe_programmable_bootstrapping.rs [hunk 8]
```diff
@@ -1406,3 +1827,101 @@ pub fn programmable_bootstrap_f128_lwe_ciphertext_mem_optimized_requirement<Scal
 ) -> Result<StackReq, SizeOverflow> {
     bootstrap_scratch_f128::<Scalar>(glwe_size, polynomial_size, fft)
 }
+
+impl<Scalar> FourierBootstrapKey<Scalar> for FourierLweBootstrapKeyOwned
+where
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_programmable_bootstrapping.rs:1832]: I think this impl should be in the entity file for now

File: tfhe/src/core_crypto/algorithms/lwe_programmable_bootstrapping.rs [hunk 9]
```diff
@@ -335,6 +480,155 @@ pub fn add_external_product_assign<Scalar, OutputGlweCont, InputGlweCont, GgswCo
...
+#[cfg_attr(__profiling, inline(never))]
+pub(crate) fn collect_next_term<Scalar: UnsignedTorus>(
+    decomposition: &mut TensorSignedDecompositionLendingIter<'_, Scalar>,
+    next_term: &mut [Scalar],
+) -> DecompositionLevel {
+    decomposition.fill_next_term(next_term).unwrap().0
+}
+
+#[cfg_attr(__profiling, inline(never))]
+pub(crate) fn update_with_fmadd(
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_programmable_bootstrapping.rs:492]: I believe the update_with_fmadd could be seen as FFT primitives ? As this allows to make polynomial mulitplications in an efficient way ? what do you think ?

File: tfhe/src/core_crypto/algorithms/lwe_programmable_bootstrapping.rs [hunk 10]
```diff
@@ -335,6 +480,155 @@ pub fn add_external_product_assign<Scalar, OutputGlweCont, InputGlweCont, GgswCo
...
+    pulp::Arch::new().dispatch(Impl {
+        output_fft_buffer,
+        lhs_polynomial_list,
+        fourier,
+        is_output_uninit,
+        fourier_poly_size,
+    })
+}
+
+pub(crate) fn update_with_fmadd_factor(
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_programmable_bootstrapping.rs:575]: same comment :)

File: tfhe/src/core_crypto/algorithms/lwe_programmable_bootstrapping.rs [hunk 11]
```diff
@@ -335,6 +480,155 @@ pub fn add_external_product_assign<Scalar, OutputGlweCont, InputGlweCont, GgswCo
     add_external_product_assign_mem_optimized(out, ggsw, glwe, fft, buffers.stack());
 }
 
+#[cfg_attr(__profiling, inline(never))]
+pub(crate) fn collect_next_term<Scalar: UnsignedTorus>(
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_programmable_bootstrapping.rs:484]: could it be a method on the TensorSignedDecomposition struct ?

File: tfhe/src/core_crypto/algorithms/lwe_programmable_bootstrapping.rs [hunk 12]
```diff
@@ -2,26 +2,158 @@
...
-use crate::core_crypto::fft_impl::fft64::math::fft::{Fft, FftView};
+use aligned_vec::CACHELINE_ALIGN;
 use concrete_fft::c64;
-use dyn_stack::{PodStack, SizeOverflow, StackReq};
+use dyn_stack::{PodStack, ReborrowMut, SizeOverflow, StackReq};
+
+use super::ggsw_conversion::convert_standard_ggsw_ciphertext_to_fourier_mem_optimized_requirement;
+
+impl<'a> FourierLweBootstrapKeyView<'a> {
+    // CastInto required for PBS modulus switch which returns a usize
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_programmable_bootstrapping.rs:33]: for now I believe this should be in the entity module

File: tfhe/src/core_crypto/algorithms/lwe_programmable_bootstrapping.rs (hunk 9 above)
Comment (sarah-ek) [reply tfhe/src/core_crypto/algorithms/lwe_programmable_bootstrapping.rs:492]: makes sense to me. i can put them with the fft

**Comment by sarah-ek** [comment]
i pushed a new commit with your suggestions. lemme know if there's anything else to fix
i'm also gonna run the benchmarks locally, to check that there are no regressions

**Comment by sarah-ek** [comment]
no regressions found on my machine

File: tfhe/src/core_crypto/entities/fourier_lwe_bootstrap_key.rs [hunk 13]
```diff
@@ -363,75 +338,3 @@ impl<'a> FourierLweBootstrapKeyView<'a> {
         );
     }
 }
-
-impl<Scalar> FourierBootstrapKey<Scalar> for FourierLweBootstrapKeyOwned
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/entities/fourier_lwe_bootstrap_key.rs:367]: Any reason to remove this implementation ?

Comment (sarah-ek) [reply tfhe/src/core_crypto/entities/fourier_lwe_bootstrap_key.rs:367]: that might have been a mistake on my part

**Comment by IceTDrinker** [comment]
closing this for now as too many things moved, we have a tracking issue for it

**Pull request closed without merging**